examples/eSealing/src/lib.rs

```diff
@@ -0,0 +1,444 @@
+    AlreadyRecorded,
+}
+
+/// Mapping the logging errors to ContractError.
+impl From<LogError> for ContractError {
+    fn from(le: LogError) -> Self {
+        match le {
+            LogError::Full => Self::LogFull,
+            LogError::Malformed => Self::LogMalformed,
+        }
+    }
+}
+
+/// The state tracked for each file.
+#[derive(Serial, DeserialWithState, Deletable, StateClone)]
+#[concordium(state_parameter = "S")]
+struct FileState<S> {
+    /// The timestamps when this file hash was recorded.
+    timestamps:    Vec<Timestamp>,
+    /// The witnesses (sender_accounts) that recorded this file hash.
+    witnesses:     Vec<AccountAddress>,
+    /// The witnesses (sender_accounts) stored in a set. This
+    /// set enables us to check that a witness has not been recorded
+    /// without iterating through the above array which
+    /// saves execution time in the `register_file` function.
+    witnesses_set: StateSet<AccountAddress, S>,
+}
+
+/// The contract state.
+#[derive(Serial, DeserialWithState, StateClone)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    files: StateMap<String, FileState<S>, S>,
+}
+
+impl<S: HasStateApi> State<S> {
+    /// Creates a new state with no files recorded.
+    fn new(state_builder: &mut StateBuilder<S>) -> Self {
+        State {
+            files: state_builder.new_map(),
+        }
+    }
+
+    /// Check if a file exists.
+    fn file_exists(&self, file_hash: &str) -> bool {
+        let record = self.files.get(&file_hash.to_string());
+
+        match record {
+            Some(_entry) => true,
+            None => false,
+        }
+    }
+
+    /// Check if a witness exists.
+    fn witness_exists(&self, file_hash: &str, witness: AccountAddress) -> bool {
+        let records = self.files.get(&file_hash.to_string());
+
+        match records {
+            Some(entry) => entry.witnesses_set.contains(&witness),
+            None => false,
+        }
+    }
+
+    /// Get all records from a specific file hash.
+    fn get_records(&self, file_hash: &str) -> ReceiveResult<(Vec<AccountAddress>, Vec<Timestamp>)> {
+        let records = self.files.get(&file_hash.to_string());
+
+        match records {
+            Some(entry) => {
+                // File exists
+                Result::Ok((
+                    entry.witnesses.iter().map(|x| *x.deref()).collect(),
+                    entry.timestamps.iter().map(|x| *x.deref()).collect(),
+                ))
+            }
+            None => Result::Ok(([].to_vec(), [].to_vec())),
+        }
+    }
+
+    /// Add a new records to a specific file hash.
+    fn add_record(&mut self, file_hash: &str, witness: AccountAddress, time: Timestamp) {
+        self.files.get_mut(&file_hash.to_string()).unwrap().timestamps.push(time);
+        self.files.get_mut(&file_hash.to_string()).unwrap().witnesses.push(witness);
+        self.files.get_mut(&file_hash.to_string()).unwrap().witnesses_set.insert(witness);
```

Comment:
Doing it like this means you are constructing 3 strings and looking up the same value in the map 3 times. It would be better to do this once.
```rust
let val = self.files.get_mut(file_hash);
if let Some(val) = val {
    val.timestamps.push(time);
    ...
}
```